{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"collapsed_sections":[],"authorship_tag":"ABX9TyN8AMaxzbr7rjuaY+Yyr73n"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","execution_count":3,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":207},"id":"r0XEpgN48-UI","executionInfo":{"status":"error","timestamp":1665554600473,"user_tz":420,"elapsed":40854,"user":{"displayName":"Dami Ibrahim","userId":"12174953129367734864"}},"outputId":"7791534e-cd7d-46f4-ede5-8b6fd90ef8b4"},"outputs":[{"output_type":"stream","name":"stdout","text":["Consensus Sequence: AAAAAGTTAAAAA\n","Best Score: 9.190159789522177\n"]},{"output_type":"stream","name":"stderr","text":["usage: ipykernel_launcher.py [options] -option1[default] <input >output\n","ipykernel_launcher.py: error: unrecognized arguments: -f /root/.local/share/jupyter/runtime/kernel-34bb2498-fb5b-44ec-a1af-8e254e347d6a.json\n"]},{"output_type":"error","ename":"SystemExit","evalue":"ignored","traceback":["An exception has occurred, use %tb to see the full traceback.\n","\u001b[0;31mSystemExit\u001b[0m\u001b[0;31m:\u001b[0m 2\n"]}],"source":["# Credit to Arpitha and Savi\n","\n","# p(psuedocounts)\n","# i(iterations)\n","# k(kmerlength)\n","\n","# inputfile\n","# >header1\n","# seq (50 bases) (6 sequences)\n","# header2\n","# seq...\n","# randomly grab a motif in the sequence depending on the input(kmerlength) hardcode input commandline\n","# put all the random motifs in a 'motif matrix'\n","# count matrix will have the counts of each base in each column\n","# always use pseudocount which adds 1 or whatever the number of pseudocounts is to each base in each column for the count matrix\n","# profile matrix will have the proportion of number of each base / the total number of all the bases.\n","# to calculate the score we need the total proportions of each sequence. If there are 300 bases in total then the Q(A) = total number of As in genome/300 = Q\n","# Then to calculate the score input the profile matrix P(X) and total proportion Q(X).\n","# The relative entropy sequence is the seq with the largest percentage of base in the sequence\n","# Get the sequence most probable based on the profile matrix\n","# To get the relative entropy sequence you get the biggest probability of each column\n","# double loop. inner loop that goes through the motif matrix to the count matrix to the profile matrix and the outer loop is to start all over.\n","# The outer loop has a specific amount of iterations.\n","# You want the smaller score when comparing scores after your done with the iterations each time.\n","# one iteration is done when the relative entropy sequence and score stops changing. After that you then have to find every possible kmer and compare\n","# probability of each base in the kmer (based on the profile matrix) and multiply them together and the highest value is the one you keep in the motif matrix.\n","# put getting the data and running the iterations in main\n","\n","\n","class CommandLine() :\n","    '''\n","    Handle the command line, usage and help requests.\n","    CommandLine uses argparse, now standard in 2.7 and beyond.\n","    it implements a standard command line argument parser with various argument options,\n","    a standard usage and help, and an error termination mechanism do-usage_and_die.\n","    attributes:\n","    all arguments received from the commandline using .add_argument will be\n","    avalable within the .args attribute of object instantiated from CommandLine.\n","    For example, if myCommandLine is an object of the class, and requiredbool was\n","    set as an option using add_argument, then myCommandLine.args.requiredbool will\n","    name that option.\n","    '''\n","\n","    def __init__(self, inOpts=None) :\n","        '''\n","        CommandLine constructor.\n","        Implements a parser to interpret the command line argv string using argparse.\n","        '''\n","        import argparse\n","        self.parser = argparse.ArgumentParser(\n","            description='Program prolog - a brief description of what this thing does',\n","            epilog='Program epilog - some other stuff you feel compelled to say',\n","            add_help=True,  # default is True\n","            prefix_chars='-',\n","            usage='%(prog)s [options] -option1[default] <input >output'\n","        )\n","        self.parser.add_argument('-i', '--iterations', nargs='?', type=int, default=1, action='store',\n","                                 help='iterations')\n","        self.parser.add_argument('-p', '--pseudocount', nargs='?', type=float, default=1, action='store',\n","                                 help='pseudocount')\n","        self.parser.add_argument('-k', '--kmer', nargs='?', type=int, default=.01, action='store',\n","                                 help='kmer length')\n","        self.parser.add_argument('-v', '--version', action='version', version='%(prog)s 0.1')\n","        if inOpts is None :\n","            self.args = self.parser.parse_args()\n","        else :\n","            self.args = self.parser.parse_args(inOpts)\n","\n","import sys\n","\n","\n","class FastAreader :\n","    '''FastAreader program provided by the professor. '''\n","\n","    def __init__(self, fname='') :\n","        '''contructor: saves attribute fname '''\n","\n","        self.fname = fname\n","        self.fileH = None\n","\n","    def doOpen(self) :\n","        if self.fname == '' :\n","            return sys.stdin\n","        else :\n","            return open(self.fname)\n","\n","    def readFasta(self) :\n","\n","        header = ''\n","        sequence = ''\n","\n","        with self.doOpen() as self.fileH :\n","\n","            header = ''\n","            sequence = ''\n","\n","            # skip to first fasta header\n","            line = self.fileH.readline()\n","            while not line.startswith('>') :\n","                line = self.fileH.readline()\n","            header = line[1 :].rstrip()\n","\n","            for line in self.fileH :\n","                if line.startswith('>') :\n","                    yield header, sequence\n","                    header = line[1 :].rstrip()\n","                    sequence = ''\n","                else :\n","                    sequence += ''.join(line.rstrip().split()).upper()\n","\n","        yield header, sequence\n","\n","import math\n","import random\n","import numpy as np\n","\n","class CRISPR:\n","    def __init__(self, iter, PC, kmers):\n","        '''\n","        Initializes the iterations, pseudocount and kmer length.\n","        Creates an empty list for the motif, count, and profile matrix.\n","        Initializes the probability for each base\n","        '''\n","        self.iter = int(iter)\n","        self.PC = int(PC)\n","        self.kmers = kmers\n","        self.motifMatrix = []\n","        self.countMatrix = []\n","        self.profileMatrix = []\n","        self.genome = []\n","        self.conSeq = None\n","        self.probability_A = 0\n","        self.probability_C = 0\n","        self.probability_T = 0\n","        self.probability_G = 0\n","\n","    def randMotif(self):\n","        '''\n","        randomly selects a motif on the genome and adds to the motif list. From Chris OH\n","        '''\n","        motifList = []\n","        for seq in self.genome:\n","            start = random.randint(0, len(seq) - self.kmers) # picks a random position in the genome to start the kmer\n","            end = int(start) + self.kmers\n","            motif = list(seq[start:end])\n","            motifList.append(motif) #add the kmer to a list\n","        self.motifMatrix = np.array(motifList) #make list an array\n","        return self.motifMatrix\n","    def getIterations(self):\n","        '''\n","        returns the number of iterations\n","        '''\n","        return self.iter\n","\n","    def arrayMotif(self, nucList):\n","        '''\n","        Converts the motif matrix into an array so I can later use the array to count the columns\n","        '''\n","        motifList = []\n","        for i in range(len(nucList)):\n","            code = list(nucList[i])\n","            motifList.append(code)\n","        self.motifMatrix = np.array(motifList)\n","        return self.motifMatrix\n","\n","    def countsMatrix(self):\n","        '''\n","        counts the number of each base in each column, plus the pseudocount and puts it in the count matrix\n","        '''\n","        A_list = []; T_list = []; C_list = []; G_list = []\n","        for val in range(self.kmers):\n","            col = self.motifMatrix[:,val] # gets the columns in the motif matrix\n","            countA = sum(np.char.count(col, 'A'), self.PC) #adds the number of bases in each column plus the pseducount\n","            countT = sum(np.char.count(col, 'T'), self.PC)\n","            countC = sum(np.char.count(col, 'C'), self.PC)\n","            countG = sum(np.char.count(col, 'G'), self.PC)\n","            A_list.append(countA)\n","            T_list.append(countT)\n","            C_list.append(countC)\n","            G_list.append(countG)\n","        count_list = [A_list, T_list, G_list, C_list] #puts the count of each base into a list\n","        self.countMatrix = np.array(count_list) #makes list an array\n","        return self.countMatrix\n","\n","    def profilesMatrix(self):\n","        '''\n","        calculates the proportion of each base in each column and puts it in the profile matrix\n","        '''\n","        A_prob = []; T_prob = []; C_prob = []; G_prob = []\n","        for val in range(self.kmers):\n","            col = self.countMatrix[:, val]\n","            x = sum(col)\n","            A_prob.append(col[0] / x) # gets the proportion of each base in each column over the total nucleotides\n","            T_prob.append(col[1] / x)\n","            C_prob.append(col[2] / x)\n","            G_prob.append(col[3] / x)\n","        prob_list = [A_prob, T_prob, C_prob, G_prob] #puts the proportion of each base into a list\n","        self.profileMatrix = np.array(prob_list) #make the list an array\n","        return (self.profileMatrix)\n","\n","    def consensusSeq(self):\n","        '''\n","        Finds the consensus sequence by iterating through the kmer length\n","        and profile matrix. Finds the max probability in each column in the profile\n","        matrix.\n","        '''\n","        seq = []\n","        for val in range(self.kmers):\n","            col = self.profileMatrix[:, val]\n","            base = max(col)\n","            if base == col[0]:  #adds the highest probability of each base to the consensus sequence\n","                seq.append('A')\n","            elif base == col[1]:\n","                seq.append('T')\n","            elif base == col[2]:\n","                seq.append('C')\n","            else:\n","                seq.append('G')\n","        self.conSeq = ''.join(seq)\n","        return self.conSeq\n","\n","    def relativeE(self):\n","        '''\n","        calculates the relative entropy score based on the profile matrix and the total number of bases in the genome.\n","        '''\n","        countRE = 0\n","        for val in range(self.kmers):\n","            col = self.profileMatrix[:, val]\n","            probA = col[0]; probT = col[1]; probC = col[2]; probG = col[3]\n","\n","            aProb = probA * math.log2(probA / self.probability_A) # calculates the entropy score for each base\n","            tProb = probT * math.log2(probT / self.probability_T)\n","            cProb = probC * math.log2(probC / self.probability_C)\n","            gProb = probG * math.log2(probG / self.probability_G)\n","            scoreRE = (aProb + tProb + cProb + gProb) # adds up all the scores\n","            countRE += scoreRE\n","        return countRE\n","\n","    def motifProb(self, seq):\n","        '''\n","        Finds the highest probability out of all the possible kmers in the sequence\n","        '''\n","        probDict = {}\n","        start = 0\n","        end = self.kmers - 1\n","        while (end < len(seq)):\n","            motifSeq = seq[start:end + 1]\n","            motifProb = 1\n","            baseNum = 0\n","            for nuc in motifSeq:\n","                if nuc == 'A':\n","                    motifProb = motifProb * self.profileMatrix[0][baseNum]\n","                elif nuc == 'T':\n","                    motifProb = motifProb * self.profileMatrix[1][baseNum]\n","                elif nuc == 'C':\n","                    motifProb = motifProb * self.profileMatrix[2][baseNum]\n","                else:\n","                    motifProb = motifProb * self.profileMatrix[3][baseNum]\n","            probDict[motifSeq] = motifProb\n","            start += 1\n","            end += 1\n","        finalMotif = max(probDict, key = lambda x: probDict[x])\n","        return finalMotif\n","\n","    def calcProb(self, sourceReader):\n","        '''\n","        Calculates every nucleotide in the file and all the nucleotides in the final to get the\n","        probability of each base.\n","        '''\n","        countA = 0; countT = 0; countC = 0; countG = 0; genomeCount = 0\n","        bases = ['A', 'T', 'C', 'G']\n","        for head, seq in sourceReader.readFasta():\n","            seq = ''.join(filter(lambda x: x in bases, seq))\n","            countA += seq.count('A')\n","            countT += seq.count('T')\n","            countC += seq.count('C')\n","            countG += seq.count('G')\n","            genomeCount += len(seq)\n","            self.genome.append(seq)\n","\n","        countA = countA + self.PC\n","        countT = countT + self.PC\n","        countC = countC + self.PC\n","        countG = countG + self.PC\n","        genomeCount = genomeCount + (4 * self.PC)\n","\n","        self.probability_A = countA / genomeCount\n","        self.probability_T = countT / genomeCount\n","        self.probability_C = countC / genomeCount\n","        self.probability_G = countG / genomeCount\n","\n","\n","def main(inFile=None, options=None):\n","    '''\n","    Gets the data runs the iterations. Check to see if the score is better or not\n","    '''\n","    cl = CommandLine(options)\n","    sourceReader = FastAreader(inFile)  # setup the Fasta reader Object\n","    thisCRISPR = CRISPR(cl.args.iterations, cl.args.pseudocount, cl.args.kmer)\n","    thisCRISPR.calcProb(sourceReader)\n","    bases = ['A', 'T', 'C', 'G']\n","\n","    bestMotif = None\n","    bestScore = math.inf\n","    for i in range(thisCRISPR.getIterations()):\n","        prevScore = -math.inf\n","        thisCRISPR.randMotif() # Get the initial set of random motifs\n","        thisCRISPR.countsMatrix()\n","        thisCRISPR.profilesMatrix()\n","        conSeq = thisCRISPR.consensusSeq()\n","        relE = thisCRISPR.relativeE()\n","        while (prevScore != relE):  #Loop through until the score stops changing\n","            var = relE\n","            prevScore = var\n","            x = []\n","            for head, seq in sourceReader.readFasta():\n","                seq = ''.join(filter(lambda x : x in bases, seq))\n","                newM = thisCRISPR.motifProb(seq)\n","                x.append(newM)\n","            thisCRISPR.arrayMotif(x)\n","            thisCRISPR.countsMatrix()\n","            thisCRISPR.profilesMatrix()\n","            conSeq = thisCRISPR.consensusSeq()\n","            relE = thisCRISPR.relativeE()\n","        if bestMotif is None and bestScore == math.inf: #if statement comparing the motif and score\n","            bestMotif = conSeq\n","            bestScore = relE\n","        else:\n","            if bestScore < relE:  # if the score stops changing then that is the new best score\n","                bestMotif = conSeq\n","                bestScore = relE\n","    print(\"Consensus Sequence: {bestMotif}\".format(bestMotif=bestMotif))\n","    print(\"Best Score: {bestScore}\".format(bestScore=bestScore))\n","\n","\n","if __name__ == \"__main__\" :\n","    main(inFile=\"input.fa\", options=[\"-i=1000\", \"-p=1\", \"-k=13\"])\n","\n","main()"]},{"cell_type":"markdown","source":["Inspections:\n","\n","Savi inspection notes:\n","        self.kmers = kmers should be int(kmers) too \n","        change defualt to type = int in command line class, should be helpful\n","        def consensusSequence(self) has if statements that check maxprobability against T twice; sub one of those if statements for G\n","        add counter to def relativeE method... important for the double summation\n","        in your motifProb, make sure you have an else statement; if you have an elif, it seems like you will have an extra nuc that is not accounted for but there is not another\n","        you don't need the lowercase nucs because FastaReader already changes it to upper()\n","\n","Jonny inspection notes:\n","  I think the algorithim in the main is good but could be more efficient by using a while and a for instead of a triple nested loop. Having a separte method for iterating through the iterations. I also think you could add more in line comments.\n","\n","Arsalan inspection notes:\n","  The profile matrix is based on the counts. He used numpy array to hold the sequences in the motif matrix. The iterations was done in the main method.\n","\n","Inspection results:\n","  I believe putting the iteration in the main method was the best and most efficient way to do so. I think next time I should add more comments to explain the code. Also some methods could probably be combined in one which could also decrease run time. My output is not completely correct and I keep getting a type error every time I run that I'm not sure how to fix. I was getting this same error last week too.\n"],"metadata":{"id":"KgJlExN6NiDU"}}]}